### PR TITLE
Add set primary domain dropdown

### PR DESCRIPTION
**Pull request by gius80**
## Proposed Changes

We noticed an increased support requests related to how set a primary domain for a site. This PR adds a new component on site domains page that should make easier to change the primary site address.

primary-1

If the site doesn't have any custom domain we show a message with links to buy and map a custom domain.

primary-2

If the site isn't eligible for setting custom domains (free plan and PWPO user) we show a message with a link to the upgrade plan page.

primary-3

## Testing Instructions

This patch should be tested in different scenarios:

**Paid plan (or free plan & exempt PWPO user) with custom domains**
- Verify that the message is `You can change it by selecting a different address from the list below. Learn more.`.
- Verify that you can change the primary domain using the dropdown.

** Paid plan (or free plan & exempt PWPO user) without custom domains**
- Verify that the message is `Before changing your primary site address you must register or connect a new custom domain.`.
- Verify that the dropdown is not rendered.

** Free plan (PWPO user) with custom domains **
** Free plan (PWPO user) without custom domains **
- Verify that the message is `Your site plan doesn't allow to set a custom domain as a primary site address. Upgrade your plan.`.
- Verify that the dropdown is not rendered.

## Pre-merge Checklist

- [X] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md for your changes?
- [X] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [X] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

### CODE REVIEW

File: client/my-sites/domains/domain-management/components/primary-domain-selector/index.tsx
Comment (Aurorum) [inline client/my-sites/domains/domain-management/components/primary-domain-selector/index.tsx]: I'm a bit confused on the logic here. If you're checking `isOnFreePlan`, won't that override PWPO users who are entitled to set a primary domain while still being on a Free plan?

File: client/my-sites/domains/domain-management/components/primary-domain-selector/index.tsx
Comment (Aurorum) [inline client/my-sites/domains/domain-management/components/primary-domain-selector/index.tsx]: What do you think about using the `InlineSupportLink` component to add a "Learn more" link to this guide? https://wordpress.com/support/domains/set-a-primary-address/

I'm not sure if users will be familiar with what a "Primary site address" actually means for them, and the effect that it will have on their site.

File: client/my-sites/domains/domain-management/components/primary-domain-selector/style.scss
Comment (Aurorum) [inline client/my-sites/domains/domain-management/components/primary-domain-selector/style.scss]: Nit: this is different on some colour schemes, and probably the more appropriate variable.

```suggestion
		color: var(--color-text);
	}
	&__subtitle {
		font-size: $font-body-small;
		margin-bottom: 1em;
		color: var(--color-text-subtle);
```

Although to be honest, I'm not even sure that the colour needs to be set for the title. 

This is also less relevant if it's changed to `FormFieldset`.

File: client/my-sites/domains/domain-management/components/primary-domain-selector/index.tsx
Comment (Aurorum) [inline client/my-sites/domains/domain-management/components/primary-domain-selector/index.tsx]: It would probably be better in terms of consistency, a11y, and saving some styles to use `FormTextInput` here and make this a `FormFieldset`.

File: client/my-sites/domains/domain-management/components/primary-domain-selector/index.tsx
Comment (Aurorum) [inline client/my-sites/domains/domain-management/components/primary-domain-selector/index.tsx]: Would it make sense to filter this out to only custom domains using `isCustomDomain`? It seems counter-intuitive to encourage users to set their primary domain as one which isn't a custom domain.

**Review by Aurorum — CHANGES_REQUESTED** [review]
This is a lovely addition! However, I think it's worth considering the use cases. 

At present, if I have a `.wpcomstaging.com` domain, I'm offered the option to set it to `.wordpress.com`. This won't work, but it appears that it does, which causes confusion to the user. 

Screenshot 2024-03-13 at 12 39 27

Likewise, it's scarily easy for me to change *away* from a custom domain, which I don't think is what we're intending here.

Screenshot 2024-03-13 at 12 41 57

I'd argue that the logic should be simplified so that this should only appear when the following conditions are true:

- The user has a custom domain linked to their site. 
- The current primary site address isn't a custom domain. 
- The user is able to set their custom domain as the primary site address.

**Comment by gius80** [comment]
Thanks for the detailed review, really appreciated!

> This is a lovely addition! However, I think it's worth considering the use cases.
> 
> At present, if I have a `.wpcomstaging.com` domain, I'm offered the option to set it to `.wordpress.com`. This won't work, but it appears that it does, which causes confusion to the user.

Nice catch! I hadn't finish to test al the cases. I filtered out `*.wordpress.com` domain on atomic sites.

> I'd argue that the logic should be simplified so that this should only appear when the following conditions are true:
> 
> * The user has a custom domain linked to their site.
> * The current primary site address isn't a custom domain.
> * The user is able to set their custom domain as the primary site address.

Those are valid points but we want to allow users to easily switch the primary domain, no matter if the domain is custom or if it's just the default one. It is true that it's easy to to change away from a custom domain but at the same time it's easy as well to set the custom domain back.

File: client/my-sites/domains/domain-management/components/primary-domain-selector/index.tsx
Comment (gius80) [reply client/my-sites/domains/domain-management/components/primary-domain-selector/index.tsx]: For PWPO users `userCanSetPrimaryDomains` will be `true`, so `shouldUpgradeToMakeDomainPrimary` will return `false`. It's the same logic we have in the table to enable the set primary address action.

File: client/my-sites/domains/domain-management/components/primary-domain-selector/index.tsx
Comment (gius80) [reply client/my-sites/domains/domain-management/components/primary-domain-selector/index.tsx]: Nice! I was considering to update the support page before adding any link but you're right, pointing to set primary address page it's a good idea.

File: client/my-sites/domains/domain-management/components/primary-domain-selector/index.tsx
Comment (gius80) [reply client/my-sites/domains/domain-management/components/primary-domain-selector/index.tsx]: See my comment https://github.com/Automattic/wp-calypso/pull/88383#issuecomment-1994990432

File: client/my-sites/domains/domain-management/components/primary-domain-selector/index.tsx
Comment (gius80) [reply client/my-sites/domains/domain-management/components/primary-domain-selector/index.tsx]: I wrapped everything inside `FormFieldset` and used `FormSettingExplanation` and `FormLabel`

File: client/my-sites/domains/domain-management/components/primary-domain-selector/index.tsx
Comment (Aurorum) [inline client/my-sites/domains/domain-management/components/primary-domain-selector/index.tsx:69]: This is what I feel is a blocker - would it make sense to hide this when `validPrimaryDomains` is only 1? The text gives the impression that there should be more, so I could imagine being easily confused here. 

Screenshot 2024-03-13 at 17 27 57

For context, this currently happens with all Simple sites without a custom domain. 

At the very least, I think that the text should be different when there's only one valid primary domain - maybe something encouraging the user to add a new domain to their site.

File: client/my-sites/domains/domain-management/components/primary-domain-selector/index.tsx
Comment (Aurorum) [inline client/my-sites/domains/domain-management/components/primary-domain-selector/index.tsx:101]: What do you think of making this an action verb so that it's a little clearer? (Not a leading question, I'm not sure about this!)

```suggestion
				{ translate( 'Change primary site address' ) }
```

If you wanted to be funky with the language here, you could even give an indication of what the primary site address refers to without needing another click. 

```suggestion
				{ translate( 'Looking to change the site address used by your viewers?' ) }
```

File: client/my-sites/domains/domain-management/components/primary-domain-selector/index.tsx
Comment (Aurorum) [inline client/my-sites/domains/domain-management/components/primary-domain-selector/index.tsx]: Another genuine question - my understanding is that `showIcon` should be true for external links like this one, and only marked false if we keep the user within Calypso when accessing the document (ie. by adding the link [here](https://github.com/Automattic/wp-calypso/blob/trunk/client/components/inline-support-link/context-links.js) and using the `supportContext` prop). 

Do you think it's better/more consistent to use `supportContext` in this case?

File: client/my-sites/domains/domain-management/components/primary-domain-selector/style.scss
Comment (Aurorum) [inline client/my-sites/domains/domain-management/components/primary-domain-selector/style.scss]: If I wanted to be incredibly nit-picky (in other words, feel free to ignore!), you could wrap the `FormFieldSet` in a `Card` and then you wouldn't need these styles.

File: client/my-sites/domains/domain-management/components/primary-domain-selector/style.scss
Comment (Aurorum) [inline client/my-sites/domains/domain-management/components/primary-domain-selector/style.scss:15]: Another nit: colour here isn't needed since it's inherited part of the component. 

```suggestion
```

File: client/my-sites/domains/domain-management/components/primary-domain-selector/index.tsx
Comment (Aurorum) [reply client/my-sites/domains/domain-management/components/primary-domain-selector/index.tsx]: Ahhh, got it! Thanks for explaining! :)

**Review by Aurorum — CHANGES_REQUESTED** [review]
Thanks for addressing the changes!

> Those are valid points but we want to allow users to easily switch the primary domain, no matter if the domain is custom or if it's just the default one. It is true that it's easy to to change away from a custom domain but at the same time it's easy as well to set the custom domain back.

I hear you. :) I still feel quite strongly that a user might make their `.wordpress.com` domain a primary address and not even realise since there's no "Save" button. Although what you're saying makes total sense, without explanation, my initial thoughts based on the prominence of this dropdown were "oh! there's something I need to action here!" 

I know that you referenced an increase in support request behinds this change, but anecdotally, I've always found that these requests are trying to make their custom domain as the primary site address. I've never had a user try to do the reverse. 

My worry is that this change would make it equally easy to do both, whereas I think it's better to nudge the user into intuitively doing what they're likely trying to achieve. 

---

But that's just my thoughts and not a blocker! Only thing which I feel is still a major issue is that this can be pretty confusing when there's no real option to set a primary domain (in which case I think there's an even stronger case for hiding it) - thoughts?

File: client/my-sites/domains/domain-management/components/primary-domain-selector/index.tsx
Comment (gius80) [reply client/my-sites/domains/domain-management/components/primary-domain-selector/index.tsx]: I think it's a good idea

**Comment by gius80** [comment]
> Thanks for addressing the changes!
> 
> > Those are valid points but we want to allow users to easily switch the primary domain, no matter if the domain is custom or if it's just the default one. It is true that it's easy to to change away from a custom domain but at the same time it's easy as well to set the custom domain back.
> 
> I hear you. :) I still feel quite strongly that a user might make their `.wordpress.com` domain a primary address and not even realise since there's no "Save" button. Although what you're saying makes total sense, without explanation, my initial thoughts based on the prominence of this dropdown were "oh! there's something I need to action here!"
> 
> I know that you referenced an increase in support request behinds this change, but anecdotally, I've always found that these requests are trying to make their custom domain as the primary site address. I've never had a user try to do the reverse.
> 
> My worry is that this change would make it equally easy to do both, whereas I think it's better to nudge the user into intuitively doing what they're likely trying to achieve.
> 
> But that's just my thoughts and not a blocker! Only thing which I feel is still a major issue is that this can be pretty confusing when there's no real option to set a primary domain (in which case I think there's an even stronger case for hiding it) - thoughts?

I discussed different option with my team and we decided to add a button to perform the change. Also we add different message on sites without custom domains, with links to registration and mapping flows.

Thanks for all the valuable suggestions!

File: client/my-sites/domains/domain-management/components/primary-domain-selector/index.tsx
Comment (Aurorum) [inline client/my-sites/domains/domain-management/components/primary-domain-selector/index.tsx]: Presumably should also be disabled when the current primary domain is the same as the selected domain?

File: client/my-sites/domains/domain-management/components/primary-domain-selector/index.tsx
Comment (Aurorum) [inline client/my-sites/domains/domain-management/components/primary-domain-selector/index.tsx]: I think that Calypso generally now prefers the term "connect" over "map" - might be wrong on this though?

Screenshot 2024-03-15 at 18 09 59

I also think the link should be on the action verbs for both cases rather than just the latter.

```suggestion
								'Before changing your primary site address, you must add a new custom domain. {{domainSearchLink}}Buy{{/domainSearchLink}} a custom domain or {{mapDomainLink}}connect{{/mapDomainLink}} a domain you already own.',
```

File: client/my-sites/domains/domain-management/components/primary-domain-selector/style.scss
Comment (Aurorum) [reply client/my-sites/domains/domain-management/components/primary-domain-selector/style.scss]: I think this probably should be made a Card now since the button effectively just makes it a Card.

**Review by Aurorum — CHANGES_REQUESTED** [review]
Thanks for taking into consideration my thoughts! :) 

The logic seems a bit off now though - even when I have a custom domain that's not the primary site address, I'm not prompted to set it as the primary site address anymore.

Screenshot 2024-03-15 at 18 07 34

(Just a note that I can't really test in-depth since I don't have any test domains to play with, but in the above case, I'm pretty sure that I should see the form)

**Comment by gius80** [comment]
Hey @Aurorum! I made a few changes that handle different scenarios (I've updated the PR description). On free plans we now show a link to upgrade the plan (only for PWPO users).

File: client/my-sites/domains/domain-management/components/primary-domain-selector/index.tsx
Comment (gius80) [reply client/my-sites/domains/domain-management/components/primary-domain-selector/index.tsx]: The current primary domain is not showed in the dropdown.

**Comment by Aurorum** [comment]
Thanks @gius80 - this is looking good, but I'm still confused on the logic. From your description, I still believe that I should be offered the option to select a primary domain in this case (have a custom domain registered, but not set). It seems to me that this is the main case which we're targeting. 

Screenshot 2024-03-18 at 17 23 21

However, I'm not offered the ability to do so.

Am I misunderstanding?! :)

File: client/my-sites/domains/domain-management/list/bulk-site-domains.tsx
Comment (leonardost) [inline client/my-sites/domains/domain-management/list/bulk-site-domains.tsx]: Nit: Can we define this function outside? I think it makes code a little more readable 🙂

File: client/my-sites/domains/domain-management/components/primary-domain-selector/index.tsx
Comment (leonardost) [inline client/my-sites/domains/domain-management/components/primary-domain-selector/index.tsx]: Is pointing to WPCOM a requirement to make a domain primary? I know it can break a user's site, but users can do that via the domain management page 🤔 here's an example where it looks weird not allowing the domain to be made primary via dropdown (that domain doesn't point to WPCOM):

Screenshot 2024-03-18 at 19 42 23

File: client/my-sites/domains/domain-management/components/primary-domain-selector/index.tsx
Comment (leonardost) [inline client/my-sites/domains/domain-management/components/primary-domain-selector/index.tsx]:
```suggestion
					{ validPrimaryDomains.length >= 1 &&
```
I think this needs to be `>= 1` because if the user has two domains and one of them is primary, the dropdown wouldn't be shown. Like in this screenshot:

Screenshot 2024-03-18 at 19 49 08

File: client/my-sites/domains/domain-management/components/primary-domain-selector/index.tsx
Comment (leonardost) [inline client/my-sites/domains/domain-management/components/primary-domain-selector/index.tsx]:
```suggestion
				{ validPrimaryDomains.length >= 1 && (
```
Same as the last comment, I think this should be `>= 1`

File: client/my-sites/domains/domain-management/components/primary-domain-selector/index.tsx
Comment (leonardost) [inline client/my-sites/domains/domain-management/components/primary-domain-selector/index.tsx]:
```suggestion
	const hasWpcomStagingDomain = domains.find(
```

I think we need to check among all domains and not only the ones that can be set as primary 🤔 otherwise this can happen:
Screenshot 2024-03-18 at 20 12 33
 
The `.wordpress.com` subdomain is not shown in the list, but it's one of the domains that's returned for the site. In this case, the `.wpcomstaging.com` subdomain is not among `validPrimaryDomains` because it's currently the primary domain. I believe this should fix the issue @Aurorum reported too.

File: client/my-sites/domains/domain-management/components/primary-domain-selector/index.tsx
Comment (leonardost) [inline client/my-sites/domains/domain-management/components/primary-domain-selector/index.tsx]: In this `onComplete` callback, I think we also need to set the `selectedDomain` state variable, otherwise this can happen:

https://github.com/Automattic/wp-calypso/assets/5324818/923dba34-3e95-4d1d-8cce-b6d9fd47dfb8

Notice that the second time I clicked on "Set as primary site address", `selectedDomain` hadn't been updated so the primary domain is always set to be the same. If I reload the page and try again, it's updated correctly.

**Review by leonardost — CHANGES_REQUESTED** [review]
This update looks great and overall is working nicely, thanks for working on this 😄 there are some issues that I think need to be addressed before merging, though. I've left comments in the code

**Comment by gius80** [comment]
> Thanks @gius80 - this is looking good, but I'm still confused on the logic. From your description, I still believe that I should be offered the option to select a primary domain in this case (have a custom domain registered, but not set). It seems to me that this is the main case which we're targeting.
> 
> Screenshot 2024-03-18 at 17 23 21
> However, I'm not offered the ability to do so.
> 
> Am I misunderstanding?! :)

You're right! I was testing a PWPO with multiple domain and din't noticed that there was an issue in the `if` condition ( `validPrimaryDomains.length > 1` instead of `> 0`). It should be working now.

File: client/my-sites/domains/domain-management/components/primary-domain-selector/index.tsx
Comment (gius80) [reply client/my-sites/domains/domain-management/components/primary-domain-selector/index.tsx]: My bad, I was testing a PWPO with many domains and didn't noticed 🤦🏻‍♂️

File: client/my-sites/domains/domain-management/components/primary-domain-selector/index.tsx
Comment (gius80) [reply client/my-sites/domains/domain-management/components/primary-domain-selector/index.tsx]: Good catch!

File: client/my-sites/domains/domain-management/components/primary-domain-selector/index.tsx
Comment (gius80) [reply client/my-sites/domains/domain-management/components/primary-domain-selector/index.tsx]: You're probably right, I set this check to be consistent with the actions in the table. I've updated it  in both places 👍

File: client/my-sites/domains/domain-management/components/primary-domain-selector/index.tsx
Comment (Aurorum) [inline client/my-sites/domains/domain-management/components/primary-domain-selector/index.tsx:206]: I'm a bit concerned about making this primary. In the following example, my custom domain is already my primary site address. Instinctively, the primary button makes me think that I should click it and make .wpstaging.com my primary address, but this probably isn't what the user wants.

Screenshot 2024-03-19 at 20 28 30

May I recommend the following?

1. Include the current primary site address in the dropdown
2. Upon loading, default the dropdown as the primary site address, especially if it's a custom domain
3. When the selected primary site address in the dropdown and the actual primary site address are the same, disable the button. 

Sorry, I know that I've got at this before - I'm just worried that this might end up achieving the opposite effect of what was intended by creating additional confusion otherwise. Let me know what you think. :)

File: client/my-sites/domains/domain-management/components/primary-domain-selector/style.scss
Comment (Aurorum) [inline client/my-sites/domains/domain-management/components/primary-domain-selector/style.scss]: `FormSelect` and `FormButton` on the same line is a bit problematic, especially responsive-wise. The strings are already a bit long, not accounting for them in different languages, so this might be quite a common problem.

| Tablet | Mobile |
|--------|--------|
| Screenshot 2024-03-19 at 20 31 03 | Screenshot 2024-03-19 at 20 30 51 | 

Perhaps better to move the button down and making it a compact button? For context, I got around this when creating a relatively similar feature like this:

Screenshot 2024-03-19 at 20 58 26

Otherwise, I think some responsive styles would be perfectly fine at resolving this. Maybe reducing the button text to "Save"? I think if the above changes were added, there would be enough context for the user to immediately know what the button refers to without a longer string.

File: client/my-sites/domains/domain-management/components/primary-domain-selector/index.tsx
Comment (Aurorum) [inline client/my-sites/domains/domain-management/components/primary-domain-selector/index.tsx]: Links are the wrong way around, I think! 

```suggestion
									domainSearchLink: (
										<a
											href={ '/domains/add/' + primaryDomain }
											onClick={ () => trackUpgradeClick( false ) }
										/>
									),
									mapDomainLink: (
										<a
											href={ '/domains/add/use-my-domain/' + primaryDomain }
											onClick={ () => trackUpgradeClick( false ) }
										/>
									),
```

**Review by Aurorum — CHANGES_REQUESTED** [review]

File: client/my-sites/domains/domain-management/components/primary-domain-selector/index.tsx
Comment (leonardost) [reply client/my-sites/domains/domain-management/components/primary-domain-selector/index.tsx]: Indeed, good catch!

File: client/my-sites/domains/domain-management/components/primary-domain-selector/index.tsx
Comment (leonardost) [reply client/my-sites/domains/domain-management/components/primary-domain-selector/index.tsx:206]: I like this suggestion 🙂 it might prevent some accidental primary address changes. What do you think @gius80?

**Review by leonardost — APPROVED** [review]
Tested again and works great, thanks for the fixes and updates! 🚀 I think this can be merged after the last comments are addressed.

File: client/my-sites/domains/domain-management/components/primary-domain-selector/index.tsx
Comment (gius80) [reply client/my-sites/domains/domain-management/components/primary-domain-selector/index.tsx]: Good catch!

File: client/my-sites/domains/domain-management/components/primary-domain-selector/index.tsx
Comment (gius80) [reply client/my-sites/domains/domain-management/components/primary-domain-selector/index.tsx:206]: @Aurorum  After an internal discussion we decided to remove the primary domain from the list and include a default label "Select a domain". To enable the button you have to select a value from the list. Considering that the explanation text already highlights the primary domain, everything should be pretty straightforward.

primary disabled

File: client/my-sites/domains/domain-management/components/primary-domain-selector/style.scss
Comment (gius80) [reply client/my-sites/domains/domain-management/components/primary-domain-selector/style.scss]: I've fixed the responsive style 👌

File: client/my-sites/domains/domain-management/components/primary-domain-selector/style.scss
Comment (Aurorum) [inline client/my-sites/domains/domain-management/components/primary-domain-selector/style.scss:32]: I'm afraid that this doesn't really work if the domain or the button is long, which could happen when the string is translated into other languages.

Screenshot 2024-03-20 at 22 20 29

What do you think of this?

```suggestion
	&__fieldset.form-fieldset {
		margin-top: 0.875rem;
		margin-bottom: 0;
		@media ( min-width: $break-large ) {
			display: flex;
		}
	}
	&__submit {
		margin: 0;
		margin-top: 12px;
		@media ( min-width: $break-large ) {
			margin: auto;
			margin-left: 12px;
		}
	}
```

File: client/my-sites/domains/domain-management/components/primary-domain-selector/index.tsx
Comment (Aurorum) [reply client/my-sites/domains/domain-management/components/primary-domain-selector/index.tsx:206]: This is much better, thank you! I think that so long as the button is initially disabled, like it is now, this is perfectly fine. 

If I had one wish for this though, I would probably add a warning notice when a user already has a custom domain and is switching away to a .wpcomstaging.com or .wordpress.com site address. I know the UI is already pretty clear, and maybe I'm underestimating our user's understanding of domains, but you'd be very surprised by how often I see this in Support! 😄 

Screenshot 2024-03-20 at 22 38 06

That is just me being extra cautious though, and possibly something to revisit depending on if there are Support tickets from confused users. :)

**Review by Aurorum — APPROVED** [review]
This is looking great and will almost certainly reduce the support tickets on primary domains! :) Thanks for addressing all my feedback - one final suggestion on styling.

**Pull request merged**
